provider/core/src/resource.rs

```diff
@@ -74,59 +73,132 @@ impl writeable::Writeable for ResourceCategory {
-        $crate::ResourceKey {
-            category: $category,
-            sub_category: $crate::internal::tinystr16!($sub_category),
-            version: $version,
+impl ResourceKey {
+    /// Gets a human-readable representation of a [`ResourceKey`].
+    /// 
+    /// The human-readable path string always contains at least one '/', and it ends with '@'
+    /// followed by one or more digits. Paths do not contain characters other than lowercase
+    /// ASCII, '_', '/', and '@'.
+    /// 
+    /// Useful for reading and writing data to a file system.
+    #[inline]
+    pub const fn get_path(&self) -> &str {
+        self.path
+    }
+
+    /// Gets a machine-readable representation of a [`ResourceKey`].
+    ///
+    /// The machine-readable hash is 4 bytes and can be used as the key in a map.
+    ///
+    /// The hash is a 32-bit FxHash of the path, computed as if on a little-endian platform.
+    #[inline]
+    pub const fn get_hash(&self) -> ResourceKeyHash {
+        self.hash
+    }
+
+    /// Creates a new ResourceKey from a path, panicking if the path is invalid.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the syntax of the path is not valid.
+    #[inline]
+    pub const fn new(path: &'static str) -> Self {
+        match Self::try_new(path) {
+            Ok(v) => v,
+            Err(_) => panic!("{}", path) // Invalid syntax!
         }
-    };
+    }
+
+    /// Creates a new ResourceKey from a path, returning an error if the path is invalid.
+    #[inline]
+    pub const fn try_new(path: &'static str) -> Result<Self, ()> {
+        match Self::check_path_syntax(path) {
+            Ok(_) => Ok(Self {
+                path,
+                _tag0: *b"ICU4XK[\x02",
+                hash: ResourceKeyHash::compute_from_str(path),
+                _tag1: *b"\x03]",
+            }),
+            Err(_) => Err(())
+        }
+    }
+
+    const fn check_path_syntax(path: &str) -> Result<(), ()> {
+        // Approximate regex: \w+(/\w+)*@\d+
+        // State 0 = start of string
+        // State 1 = after first character
+        // State 2 = after a slash
+        // State 3 = after a character after a slash
+        // State 4 = after @
+        // State 5 = after a digit after @
+        let mut i = 0;
+        let mut state = 0;
+        let path_bytes = path.as_bytes();
+        while i < path_bytes.len() {
+            let c = path_bytes[i];
+            state = match (state, c) {
+                (0 | 1, b'a'..=b'z' | b'0'..=b'9' | b'_') => 1,
+                (1, b'/') => 2,
+                (2 | 3, b'a'..=b'z' | b'0'..=b'9' | b'_') => 3,
+                (3, b'/') => 2,
+                (3, b'@') => 4,
+                (4 | 5, b'0'..=b'9') => 5,
+                _ => return Err(())
+            };
+            i += 1;
+        }
+        if state != 5 {
+            return Err(())
+        }
+        Ok(())
+    }
+}
+
+#[test]
+fn test_path_syntax() {
```

Comment:
Nit: so far I've always seen tests in a `mod test` at the bottom of the file. Does the style guide have an opinion on this?

Reply:
I put it here to be closer to the code that it is testing.  I don't think the style guide is very opinionated about where tests go.  It often makes sense to put them at the bottom of the file, but not always?